### PR TITLE
Rework to support scaling under Kubernetes/OpenShift

### PR DESCRIPTION
**Pull request by mfojtik**
Hi guys,

This is a partial rework of this image to have ZooKeeper running on OpenShift/Kubernetes in a way where you can scale/scale down it using replicationController.

To try this you can:

```
$ docker build -t zookeeper .
$ osc create -f kubernetes/zookeeper_config.json
```

Assuming that you have OpenShift running you should see following:

```
$ osc get pods
POD               IP            CONTAINER(S)   IMAGE(S)    HOST                           LABELS                  STATUS    CREATED
zookeeper-577sj   172.17.0.40   server         zookeeper   openshiftdev.local/127.0.0.1   name=zookeeper-server   Running   9 seconds
zookeeper-a2h8z   172.17.0.39   server         zookeeper   openshiftdev.local/127.0.0.1   name=zookeeper-server   Running   9 seconds
zookeeper-dd2by   172.17.0.38   server         zookeeper   openshiftdev.local/127.0.0.1   name=zookeeper-server   Running   9 seconds
```

To scale it up, you can increase the number of 'replicas':

```
$ osc update rc zookeeper --patch='{ "apiVersion": "v1beta1", "desiredState": { "replicas": 4 }}
```

FYI, this is a work in progress stuff, README is missing and there are couple of failover scenarios/persistent storage scenarios that are not covered (yet).

PTAL

### CODE REVIEW

**Comment by iocanel** [comment]
This looks really really awesome!

I do have some questions though:

When scale up/down happens, I see that zookeeper services are restarted, right?.
The problem I see, is that if we restart zookeeper servers without some short of coordination, data loss is possible.

Personally, I think that its really hard to relibly resize a zookeeper ensemble in the 3.4.x version of zookeeper (due to zookeeper limitations) and maybe we should wait for 3.5.0 (which is currently alpha) that provides an api for adding and removing servers.

Thoughts?

**Comment by mfojtik** [comment]
@iocanel 

Yes, the server is restarted to pickup the new configuration. I was thinking to put some random sleep to minimize the risk of data loss (so the servers will not restart at the same moment).

I was also playing with 3.5.0-alpha but I had no luck getting 'reconfigure' work :( It prints that reconfiguring is not implemented (or something similar). Do you have any insight to when they will release 3.5.0?

**Comment by mfojtik** [comment]
CC / @danmcp @smarterclayton

**Comment by iocanel** [comment]
AFAIR its a bit more complicated than this. You don't just need to guarantee that there will be quorum. You also need to ensure that at any given time the leader has the latest stuff (random sleeps don't help here).

**Comment by mfojtik** [comment]
@iocanel @danmcp I reworked this PR to have static list of 3 zookeeper servers wired together with 3 services. This won't scale up or down, but work as a static unit. PTAL.

File: example/zookeeper_config.json [hunk 1]
```diff
@@ -0,0 +1,202 @@
+{
+  "kind": "Config",
```
Comment (jimmidyson) [inline example/zookeeper_config.json]: Is there a reason to use `Config` vs `List`?

Comment (smarterclayton) [inline example/zookeeper_config.json]: Everyone should be using List

----- Original Message -----

Comment (mfojtik) [inline example/zookeeper_config.json]: +1 old habits...

**Comment by smarterclayton** [comment]
When we have persistent volumes the volume will give us the necessary lock to ensure two pods with the same identity aren't running under the same service.

**Comment by iocanel** [comment]
@mfojtik with a very quick glance it looks good to me. I'll look closer tomorrow first thing in the morning.

**Comment by iocanel** [comment]
"First thing in the morning?" ... Apologies for the long delay!
I cherry-picked your commit into the master!

Thanks again for it, it simplifies things I great deal and I love it :-)

**Pull request closed without merging**